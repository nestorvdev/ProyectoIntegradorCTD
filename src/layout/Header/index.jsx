import React, { useEffect, useState } from 'react';
import logo from "./img/logoWguest.jpg";
import Styles from "./styles.module.css"
import StylesApp from "../../App.module.css"
import StylesLayout from "../styles.module.css"
import MenuBurger from '../../components/MenuBurger/MenuBurger';
import MenuButton from '../../components/MenuBurger/MenuButton';
import { Link } from "react-router-dom";
import lineOrange from "./img/LineOrange.png"
import verticalLine from "./img/VerticalLine.png"

export default function Header({ setLastLocation, setBookingWithoutLogin, setLoading, activeCreate, activeLogin, isLogged, showBurger, setShowBurger, handleClean, handleFavourite }) {
    const role = sessionStorage.getItem("role");
    const [admin, setAdmin] = useState(false);
    const [adminMenu, setAdminMenu] = useState(false);
    const [userMenu, setUserMenu] = useState(false);
    const showUserName = (isLogged) ? `${Styles.user} ${Styles.loggedIn}` : Styles.user;
    const hideButtons = (isLogged) ? `${Styles.buttons} ${Styles.user}` : Styles.buttons;
    const administration = isLogged ? Styles.optionsBox : Styles.hideButton;

    /* const baseUrl = "http://localhost:8080/" */

    function handleLogOut() {
        setLoading(true)
        sessionStorage.removeItem("id")
        sessionStorage.removeItem("email")
        sessionStorage.removeItem("token")
        sessionStorage.removeItem("role")
        sessionStorage.removeItem("startDate")
        sessionStorage.removeItem("endDate")
        sessionStorage.setItem("log", "false")
        setLastLocation(window.location.pathname)
        document.querySelectorAll("nav")[1].classList.remove(`${Styles.openNav}`)
        document.querySelectorAll("nav")[1].classList.add(`${Styles.closeNav}`)
    }

    function handleFavouriteClick() {
        handleFavourite()
        handleShowNav()
    }

    function handleShow() {
        setShowBurger(true)
    }
    function handleHide() {
        setShowBurger(false)
        setBookingWithoutLogin(false)
    }

    function handleErrorLogin() {
        setBookingWithoutLogin(false)
    }

    function handleShowNav() {
        setUserMenu(true);
        setAdminMenu(false);
    }

    function handleCloseNav() {
        setUserMenu(false);
        setAdminMenu(false);
    }

    let handleAdminMenu = () => {
        setAdminMenu(true);
        setUserMenu(false);
    }

    useEffect(() => {
        if (sessionStorage.getItem("role") === "ADMIN") {
            setAdmin(true)
        } else {
            setAdmin(false)
        }
    }, [role])


    return (
        <div className={Styles.containerHeader}>
            <header className={`${Styles.header} ${StylesApp.delimiter}`} onMouseLeave={handleCloseNav}>
                <div className={showBurger === true ? `${Styles.headerTop} ${StylesLayout.opacity} ${StylesApp.delimiterChild}` : `${Styles.headerTop} ${StylesApp.delimiterChild}`}>
                    <Link to="/" className={Styles.home} onClick={handleHide}>
                        <div className={Styles.logo} onClick={handleClean}>
                            <img src={logo} alt="logo" />
                            <h3>Un mundo por descubrir</h3>
                        </div>
                    </Link>
                    <div className={hideButtons}>
                        <Link to="/create" >
                            <button id="createC" className={activeCreate ? Styles.hideButton : null} >
                                Crear Cuenta
                            </button>
                        </Link>
                        <Link to="/login">
                            <button className={activeLogin ? Styles.hideButton : null} onClick={handleErrorLogin} >
                                Iniciar Sesión
                            </button>
                        </Link>
                    </div>

                    <div className={administration}>
                        {admin &&
                            <div className={administration} onMouseMove={handleAdminMenu}>
                                <>
                                    <div className={Styles.boxAdminArrow}>
                                        <h4 className={Styles.administracion}>Administración</h4>
                                        <span className={Styles.arrowAdmin}></span>
                                    </div>
                                    <img className={Styles.verticalLine} src={verticalLine} alt="line" />
                                </>
                            </div>
                        }

                        <div className={showUserName} onMouseMove={handleShowNav} >
                            <div>
                                <div className={Styles.logoName}>
                                    <p>{sessionStorage.getItem("iniciales")}</p>
                                </div>
                                <div className={Styles.text}>
                                    <h3 className={Styles.great}>Hola,</h3>
                                    <h3 className={Styles.name}>{sessionStorage.getItem("name")} {sessionStorage.getItem("surname")}</h3>
                                </div>
                                <span className={Styles.arrow}></span>
                            </div>
                        </div>
                    </div>
                    <MenuButton show={showBurger} handleShow={handleShow} />
                </div>
                <MenuBurger admin={admin} handleFavouriteClick={handleFavouriteClick} show={showBurger} handleHide={handleHide} isLogged={isLogged} iniciales={sessionStorage.getItem("iniciales")} activeLogin={activeLogin} activeCreate={activeCreate} handleLogOut={handleLogOut} handleFavourite={handleFavourite} />
                <div className={userMenu ? `${Styles.headerOptions} ${Styles.userOptions}` : Styles.hideButton}/* className={Styles.closeNav} */ /* onMouseLeave={handleCloseNav} */>
                    <Link to="/">
                        <h4 className={Styles.seeMyAccount} onClick={handleFavouriteClick}>Ver favoritos</h4>
                    </Link>
                    <img src={lineOrange} alt="" />
                    <Link to="/mybookings">
                        <h4 className={Styles.seeMyAccount}>Mis reservas</h4>
                    </Link>
                    <img src={lineOrange} alt="" />
                    <h4 className={Styles.seeMyAccount}><a href="/" onClick={handleLogOut}>Cerrar sesión</a> </h4>
                </div>
                <div className={adminMenu ? `${Styles.headerOptions} ${Styles.adminOptions}` : Styles.hideButton} >
                    <Link to="/product/create">
                        <h4 className={Styles.seeMyAccount}>Crear Producto</h4>
                    </Link>
                    <img src={lineOrange} alt="" />
                    <Link to="/product/update">
                        <h4 className={Styles.seeMyAccount}>Modificar Producto</h4>
                    </Link>
                </div>
            </header>

        </div >
    )
}