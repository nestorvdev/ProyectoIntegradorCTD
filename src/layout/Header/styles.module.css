header.header {
  background-color: white;
  display: flex;
  flex-wrap: nowrap;
  height: 80px;
  justify-content: center;
  padding-top: 5px;
  position: fixed; 
  top:0;
  width: 100%;
  z-index: 100;
}

.headerTop{
  align-items: center;
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  padding-left: clamp(10px,3.9%,40px);
  padding-right: clamp(10px,3.9%,40px);
  width: 100%;
  z-index: 101;
}

/* css contenedor logo  */
header.header a {
  text-decoration: none;
}

div.buttons>a:first-child{
  margin-right: 10px;
}

.home div.logo {
  align-items: flex-start;
  display: flex;
  flex-direction: column;
}

.home div.logo img {    
  width: 210px;
}

.home div.logo h3 {
  color: var(--tertiary-color);
  display: flex;
  font-size: var(--heading4);
  font-weight: var(--font-regular);    
  width: 270px;
}
  
/* css contenedor botones  */

div.buttons {
  display: flex;
  padding-right: 8px;
}

div.buttons button {
  background-color: white;
  border-radius: 5px;
  border: 1px solid var(--primary-color);
  color: var(--primary-color);
  font-size: var(--button1);
  font-weight: var(--font-bold);
  padding: 5pt;
  width: clamp(164px,22.2vw,206px);
}

  
div.buttons button:hover {
  background-color: var(--primary-color);
  color: white;
  cursor: pointer;  
}

.hideButton{
  display:none;
}

/* css contenedor user  */

div.user {
  display: none; 
}

/*Clase para que se vea el usuario en el header cuando inicia sesion*/
div.loggedIn {
  display: flex;
  flex-direction: column;
}

div.loggedIn>div {
  display: flex;
}

.containerHeader{
  position: relative;
}

/* .containerHeader>nav{
  background-color: white;
  border:solid 2px black;
  border-top: none;
  margin-right: clamp(10px, 3.9%, 40px);
  padding: 10px;
  position: absolute;
  right: calc((100vw - 1366px) / 2);
  transition: all 1s;
  width: 210px;
  z-index: 50;   
    
} */

/* .containerHeader>nav a{
  color:black;
}

.containerHeader>nav img{
  width: 190px;
} */
  
/* .openNav{
  top:80px;  
}

.closeNav{
  top: -180px;  
} */

.arrow{
  border-style: solid;
  border-width: 0 2px 2px 0;
  content: '';
  display: inline-block;
  height: 8px;
  position: relative;
  transform: rotate(45deg);
  width: 8px;
  opacity: 0.5;
  margin: 16px 0 0 6px;
  vertical-align: top;
}

div.user div {
  padding: 0 5px;
}

div.user div.logoName {
  align-items: center;
  display: flex;
  flex-wrap: wrap;
}

div.user div.logoName p {
  background-color: var(--secondary-color);
  border-radius: 50%;
  color: white;
  font-size: var(--heading2);
  font-weight: var(--font-bold);
  padding: 5px;
}
  

div.user div.text h3 {
  font-size: var(--heading3);
}

div.user div.text h3.great {
  opacity: 0.5;
}

div.user div.text h3.name {
  color: var(--primary-color);
}

div.user div.text h4.seeMyAccount {     
  z-index : 50;   
}

div.user div.text h4.seeMyAccount:hover {
  cursor:pointer;
  color:var(--primary-color);  
  background-color:aquamarine; 
}

div.user div.close{
  font-size: 22px;
  font-weight: var(--font-bold);
  position: relative;
  right: 20px;
  top: -10px;
}

.optionsBox {
  display: flex;
  flex-direction: row;
  align-items: center;
}

.optionsBox .verticalLine {
  padding: 0px 18px;
}

.administracion {
  text-decoration: none;
  color: var(--secondary-color);
  padding-right: 5px;
}

.arrowAdmin{
  border-style: solid;
  border-width: 0 2px 2px 0;
  content: '';
  display: inline-block;
  height: 7px;
  position: relative;
  transform: rotate(45deg);
  width: 7px;
  opacity: 0.5;
  margin-left: 6px;
  vertical-align: top;
}

.headerOptions {
  background-color: white;
  border: solid 2px black;
  border-top: none;
  margin-right: clamp(10px, 3.9%, 40px);
  padding: 0 10px;
  position: fixed;   
  z-index: 49;
  display: flex;
  flex-direction: column;  
  animation: fadeIn 0.5s;
}

@keyframes fadeIn {
  from {
    top: 5%;
    opacity: 0.2;
  }
  to {
    top: 80px;
    opacity: 1;
  }
}

.adminOptions {     
  right: calc((100vw - 883px) / 2);
  top: 80px; 
  width: 174px;  
}

.userOptions {
  right: calc((100vw - 1366px) / 2);
  top: 80px; 
  width: 215px;  
}

.adminOptions a, .userOptions a{
  color:black;
}

.adminOptions img{
  width: 98%;  
}

.userOptions img{
  width: 98%;  
}

.seeMyAccount {
  padding: 10px 0;
  margin: 10px 0;  
}

.seeMyAccount:hover{
  cursor: pointer;
  background-color:#f0d7d1;    
}

.boxAdminArrow {
  display: flex;
  align-items: center;
}

/* css contenedor mobile  */

div.mobile {
  display: none; /* cambiar a block para mostrarlo */
  justify-content: center;
  margin-left: 20px;
}


  
/* media queries */
@media all and (max-width: 1366px) {
  .userOptions{
    left: calc(100vw - 290px);
  }

  .adminOptions{
    left: calc(100vw - 470px);
  }
}
  
@media all and (max-width: 769px) {
  /* Header */
  a.home div.logo h3 {
    display: none;
  }

  .home div.logo img {    
    width: 200px;
    
  }

  div.buttons button {
    width: 144px;
  }
  
  header.header{
    height: 99px;
    padding: 5px 4px;
  }

  .headerTop{
    height: 100%;
  }

  .administracion{
    padding-top:6px;
  }

  div.user div.text h3 {
    font-size: 12pt;
  }

  div.user div.logoName p {
    font-size: var(--heading4);
  }

  div.user div {
    padding: 0 2px;
  }

  div.user div.text h4.seeMyAccount {
    font-size: 10pt;
  }

  @keyframes fadeIn {
    from {
      top: 5%;
      opacity: 0.2;
    }
    to {
      top: 99px;
      opacity: 1;
    }
  }
  
  .userOptions{
    left: calc(100vw - 214px);
    top:99px;
    width: 175px;
  }

  .adminOptions{
    left: calc(100vw - 400px);
    top:99px;
  }
}

@media all and (max-width: 600px) {
  .optionsBox{
    flex-direction: column;
    flex-flow: column-reverse;   
  }

  

  .verticalLine{
    display: none;
  }

  .userOptions{
    left: calc(100vw - 210px);
  }
  .adminOptions{
    left: calc(100vw - 210px);
  }
}

@media all and (max-width: 415px) {
  /* Header */

  header.header {
    box-shadow: 0 4px 4px rgba(0, 0, 0, 0.3);
    height: 93px;
    padding: 5px 11px;
  } 
  

  div.buttons {
    display: none;
  }

  div.mobile {
    display: block;
  }

  div.loggedIn{
    display: none;
  }

  div.botones>a>p:hover{
    background-color: var(--primary-color);
    border-radius: 5px;
    color: white;
  }

  .optionsBox {
    display: none;
  }
}