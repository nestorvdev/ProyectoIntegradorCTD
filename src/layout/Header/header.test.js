import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import Header from './index';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

describe("Probando el componente <Header/>", () => {
    let wrapper;
    let setActiveCreate = jest.fn(); //Simula la funcion del set
    let setActiveLogin = jest.fn() //Simula la funcion del set
    let handleClean = jest.fn()
    let handleFavourite = jest.fn()
   
    beforeEach(() => {
        wrapper = shallow(<Header setActiveCreate={setActiveCreate} setActiveLogin={setActiveLogin} handleClean={handleClean} handleFavourite ={handleFavourite}/>)
    });

    it("Deberia mostrar <Header/> correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("Deberia retornar 3 etiquetas h3", () => {
        let node = wrapper.find("h3")
        expect(node.length).toEqual(3)
    });

    it("Deberia retornar 3 Links", () => {
        let node = wrapper.find("Link")
        expect(node.length).toEqual(3)
    });

    it("Deberia retornar 2 botones", () => {
        let node = wrapper.find("button")
        expect(node.length).toEqual(2)
    });

    it("Deberia llamar a handleClean cuando se hace click en el logo", ()=>{
        wrapper.find("div.logo").simulate("click")
        expect(handleClean).toBeCalled()
    })

    it("Deberia llamar a handleClean cuando se hace click en el logo", ()=>{
        wrapper.find("h4.seeFavourite").simulate("click")
        expect(handleFavourite).toBeCalled()
    })
});

