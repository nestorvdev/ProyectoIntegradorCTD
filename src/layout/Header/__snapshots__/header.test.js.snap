// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente <Header/> Deberia mostrar <Header/> correctamente 1`] = `
<div
  className="containerHeader"
>
  <header
    className="header delimiter"
    onMouseLeave={[Function]}
  >
    <div
      className="headerTop delimiterChild"
    >
      <Link
        className="home"
        onClick={[Function]}
        to="/"
      >
        <div
          className="logo"
          onClick={[MockFunction]}
        >
          <img
            alt="logo"
            src="logoWguest.jpg"
          />
          <h3>
            Un mundo por descubrir
          </h3>
        </div>
      </Link>
      <div
        className="buttons"
      >
        <Link
          to="/create"
        >
          <button
            className={null}
            id="createC"
          >
            Crear Cuenta
          </button>
        </Link>
        <Link
          to="/login"
        >
          <button
            className={null}
            onClick={[Function]}
          >
            Iniciar Sesión
          </button>
        </Link>
      </div>
      <div
        className="hideButton"
      >
        <div
          className="user"
          onMouseMove={[Function]}
        >
          <div>
            <div
              className="logoName"
            >
              <p />
            </div>
            <div
              className="text"
            >
              <h3
                className="great"
              >
                Hola,
              </h3>
              <h3
                className="name"
              >
                 
              </h3>
            </div>
            <span
              className="arrow"
            />
          </div>
        </div>
      </div>
      <MenuButton
        handleShow={[Function]}
      />
    </div>
    <MenuMobile
      admin={false}
      handleFavourite={[MockFunction]}
      handleFavouriteClick={[Function]}
      handleHide={[Function]}
      handleLogOut={[Function]}
      iniciales={null}
    />
    <div
      className="hideButton"
    >
      <Link
        to="/"
      >
        <h4
          className="seeMyAccount"
          onClick={[Function]}
        >
          Ver favoritos
        </h4>
      </Link>
      <img
        alt=""
        src="LineOrange.png"
      />
      <Link
        to="/mybookings"
      >
        <h4
          className="seeMyAccount"
        >
          Mis reservas
        </h4>
      </Link>
      <img
        alt=""
        src="LineOrange.png"
      />
      <h4
        className="seeMyAccount"
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Cerrar sesión
        </a>
         
      </h4>
    </div>
    <div
      className="hideButton"
    >
      <Link
        to="/product/create"
      >
        <h4
          className="seeMyAccount"
        >
          Crear Producto
        </h4>
      </Link>
      <img
        alt=""
        src="LineOrange.png"
      />
      <Link
        to="/product/update"
      >
        <h4
          className="seeMyAccount"
        >
          Modificar Producto
        </h4>
      </Link>
    </div>
  </header>
</div>
`;
